src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java

```diff
@@ -118,25 +118,21 @@ String getFullPath(String dirPath, String name) {
      * Our objects are small, so I think we can use the maximum batch size without
      * concern for using too much memory.
      */
-    void scanCollectionObjects(Firestore firestore,
+    <V> void scanCollectionObjects(Firestore firestore,
```

Comment:
There are two uses of this method. It appears you did not change the other one to use ApiFutureGenerator.
I'm not sure you should. This is radically different than what is needed for the simple delete case.

Reply:
I didn't actually have to since `ApiFutureGenerator` is a functional interface.  I was able to leave the other call alone.  I get your point about the different uses.  That being said, in this case, I don't think we can avoid the threadpool since the real bottleneck appears to be the GCS deletes (not surprisingly)

Reply:
First off, this is a short-term problem. We should change our algorithm for deleting a bucket to use the Google best practices at the point when Stairway has long wait support. Then we do not visit all of the files.

> [Side bar: that operation needs to be done out of band of the user's delete. One thing to do is to have the delete flight launch a separate cleaning flight that does the bucket delete followed by the collection delete.]

In any case, I think we should keep that as a separately replaceable operation.

Second, in the short-term, file delete is not something to over optimize. It is not a key performance issue.
I think we should  write one, batching, retrying, simple-no-function-inputs method to delete all entries in a collection that we can use for files, directories, and dependencies.

Separately, we can use the thread pool to delete the files. It would be good hygiene to delete the files first, since then we still have the firestore collection to drive a retry of the file deletes if there are failures.

Reply:
My thinking with optimizing is that it's reaaaaally painful for testing.  Like I said, for 10K files, it takes almost an hour and one day, someone will want to do a delete of a bunch of files and it will take potentially days to run.

Reply:
I agree that we need to optimize. What I was trying to suggest is: we use threadpool to parallelize the file deletes, and then use the firestore batching to delete the collection. It is not as optimized as doing one pass and deleting both, but I'm arguing that is tolerable. Especially since we know we will be changing the bucket delete algorithm in the future.
Hope that makes more sense! We can discuss tomorrow.